internal/response-consumer/handler.go

```diff
@@ -172,6 +163,58 @@ func (this *handler) onMessage(ctx context.Context, msg *k.Message) {
 	}
 }
 
+func createRecord(ctx context.Context, tx *gorm.DB, toCreate []db.RunHost) error {
+	createResult := tx.Model(db.RunHost{}).
+		Clauses(clause.OnConflict{
+			Columns:   []clause.Column{{Name: "run_id"}, {Name: "host"}},
+			DoUpdates: clause.AssignmentColumns([]string{"status", "log"}),
+		}).
+		Create(&toCreate)
+
+	if createResult.Error != nil {
+		utils.GetLogFromContext(ctx).Errorw("Error upserting run hosts in db", "error", createResult.Error)
+		return createResult.Error
+	}
+
+	return nil
+}
+
+func hostToUpdate(ctx context.Context, tx *gorm.DB, runId uuid.UUID, hostName string) *db.RunHost {
+	host := db.RunHost{}
+	baseQuery := tx.Model(db.RunHost{}).Where("run_id=?", runId).Where("host=?", hostName)
+
+	if selectResult := baseQuery.Order("sat_sequence desc").First(&host); selectResult.Error != nil {
+		if errors.Is(selectResult.Error, gorm.ErrRecordNotFound) {
+			return nil
+		}
+		utils.GetLogFromContext(ctx).Errorw("Error fetching runHost from db", "error", selectResult.Error)
+	}
+	return &host
+}
+
+func createUpdateRecord(ctx context.Context, tx *gorm.DB, responseFull bool, toUpdate []db.RunHost) error {
+	if responseFull != true {
+		for _, runHost := range toUpdate {
+			if host := hostToUpdate(ctx, tx, runHost.RunID, runHost.Host); host != nil {
+				if *host.SatSequence != *runHost.SatSequence-1 {
+					runHost.Log = "\n...\n" + runHost.Log
+				}
+
+				runHost.Log = host.Log + runHost.Log
```

Comment:
This can be done better:
a) more atomically by implementing the concatenation in sql as opposed to read+write (in theory the value in database may get updated in between the read/write by a concurrent consumer)
b) validate that the received sequence is not <= the last stored sequence
c) there is a special unicode character for ellipsis https://en.wikipedia.org/wiki/Ellipsis
d) we should be careful about the number of SQL queries needed to process a single run_host - reduce if possible

Reply:
I was able to accomplish this with a single query with the help of sql `CASE` statements. Unfortunately, gorm doesn't support case statements out of the box, therefore I had to embed a portion of sql into the gorm interface.
I tried to keep the `assignmentWithCase()` as simple as possible. Lmk, if you see any room for improvements there.

> b) validate that the received sequence is not <= the last stored sequence

Right now, when we receive sequence <= the last stored sequence, we don't update the `sat_sequence` number and the data loss indicator in `log` column isn't  being added.